pkg/resources/catalog.cattle.io/v1/clusterrepo/validator.go

```
@@ -0,0 +1,103 @@
+	admitter admitter
+}
+
+// GVR returns the GroupVersionKind for this CRD.
+func (v *Validator) GVR() schema.GroupVersionResource {
+	return gvr
+}
+
+// Operations returns list of operations handled by this validator.
+func (v *Validator) Operations() []admissionregistrationv1.OperationType {
+	return []admissionregistrationv1.OperationType{
+		admissionregistrationv1.Update,
+		admissionregistrationv1.Create,
+	}
+}
+
+// ValidatingWebhook returns the ValidatingWebhook used for this CRD.
+func (v *Validator) ValidatingWebhook(clientConfig admissionregistrationv1.WebhookClientConfig) []admissionregistrationv1.ValidatingWebhook {
+	return []admissionregistrationv1.ValidatingWebhook{*admission.NewDefaultValidatingWebhook(v, clientConfig, admissionregistrationv1.NamespacedScope, v.Operations())}
+}
+
+// Admitters returns the admitter objects used to validate clusterrepos.
+func (v *Validator) Admitters() []admission.Admitter {
+	return []admission.Admitter{&v.admitter}
+}
+
+type admitter struct {
+}
+
+// Admit is the entrypoint for the validator. Admit will return an error if it is unable to process the request.
+// If this function is called without NewValidator(..) calls will panic.
+func (a *admitter) Admit(request *admission.Request) (*admissionv1.AdmissionResponse, error) {
+	if request.Operation == admissionv1.Create || request.Operation == admissionv1.Update {
+		newClusterRepo, err := v1.ClusterRepoFromRequest(&request.AdmissionRequest)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get clusterRepo from request: %w", err)
+		}
+
+		var fieldErr *field.Error
+		if err := a.validateFields(newClusterRepo); err != nil {
+			if errors.As(err, &fieldErr) {
+				return admission.ResponseBadRequest(fieldErr.Error()), nil
+			}
+			return nil, fmt.Errorf("failed to validate fields on ClusterRepo: %w", err)
+		}
+	}
+
+	return admission.ResponseAllowed(), nil
+}
+
+func (a *admitter) validateFields(newClusterrepo *catalogv1.ClusterRepo) *field.Error {
```

This method should return a regular error in the signature. And it's in the caller, `Admit`, that you inspect the actual type of the error.

---

Done.